cmd/searcher/internal/search/matchtree.go

```diff
@@ -231,9 +233,143 @@ func (rm *regexMatcher) ToZoektQuery(matchContent bool, matchPath bool) (zoektqu
 		}), nil
 }
 
-func (rm *regexMatcher) negateIfNeeded(q zoektquery.Q) zoektquery.Q {
+func (rm *regexMatchTree) negateIfNeeded(q zoektquery.Q) zoektquery.Q {
 	if rm.isNegated {
 		return &zoektquery.Not{Child: q}
 	}
 	return q
 }
+
+type andMatchTree struct {
+	children []matchTree
+}
+
+func (am *andMatchTree) MatchesString(s string) bool {
+	for _, m := range am.children {
+		if !m.MatchesString(s) {
+			return false
+		}
+	}
+	return true
+}
+
+func (am *andMatchTree) MatchesFile(fileBuf []byte, limit int) (bool, [][]int) {
+	var matches [][]int
+	for _, m := range am.children {
+		// Pass the full limit to the children instead of tracking how many matches we
+		// have left. This is slightly wasteful but keeps the logic simpler.
+		childMatch, childMatches := m.MatchesFile(fileBuf, limit)
+		if !childMatch {
+			return false, nil
+		}
+		matches = append(matches, childMatches...)
+	}
+
+	return true, mergeMatches(matches, limit)
+}
+
+func (am *andMatchTree) ToZoektQuery(matchContent bool, matchPath bool) (zoektquery.Q, error) {
+	var children []zoektquery.Q
+	for _, m := range am.children {
+		q, err := m.ToZoektQuery(matchContent, matchPath)
+		if err != nil {
+			return nil, err
+		}
+		children = append(children, q)
+	}
+	return &zoektquery.And{Children: children}, nil
+}
+
+func (am *andMatchTree) String() string {
+	return fmt.Sprintf("AND (%d children)", len(am.children))
+}
+
+type orMatchTree struct {
+	children []matchTree
+}
+
+func (om *orMatchTree) MatchesString(s string) bool {
+	for _, m := range om.children {
+		if m.MatchesString(s) {
+			return true
+		}
+	}
+	return false
+}
+
+func (om *orMatchTree) MatchesFile(fileBuf []byte, limit int) (bool, [][]int) {
+	match := false
+	var matches [][]int
+	for _, m := range om.children {
+		// Pass the full limit to the children instead of tracking how many matches we
+		// have left. This is slightly wasteful but keeps the logic simpler.
+		childMatch, childMatches := m.MatchesFile(fileBuf, limit)
+		match = match || childMatch
+		matches = append(matches, childMatches...)
+	}
+
+	return match, mergeMatches(matches, limit)
+}
+
+func (om *orMatchTree) ToZoektQuery(matchContent bool, matchPath bool) (zoektquery.Q, error) {
+	var children []zoektquery.Q
+	for _, m := range om.children {
+		q, err := m.ToZoektQuery(matchContent, matchPath)
+		if err != nil {
+			return nil, err
+		}
+		children = append(children, q)
+	}
+	return &zoektquery.Or{Children: children}, nil
+}
+
+func (om *orMatchTree) String() string {
+	return fmt.Sprintf("OR (%d children)", len(om.children))
+}
+
+// mergeMatches sorts the matched ranges, merges overlapping ranges,
+// and finally truncates the list to obey 'limit'
+func mergeMatches(matches [][]int, limit int) [][]int {
+	if len(matches) == 0 {
+		return matches
+	}
+
+	sort.Sort(matchSlice(matches))
+	newMatches := make([][]int, 0, len(matches))
```

Comment:
very minor: we can avoid the allocation here and do the merging in-place
```suggestion
	newMatches := matches[:0]
```

Reply:
Oops, I meant to do this but forgot! I was so excited to use this new Go trick I learned.